### PR TITLE
filebeat/input/udp: add initial UDP metrics support

### PR DESCRIPTION
**Pull request by efd6**
## What does this PR do?

This adds metrics for UDP packet count and total bytes, and histograms for time required to process UDP packets prior to acking from a publication and time between UDP packet arrivals.

## Why is it important?

This allows us to help users configure their systems to match the requirements that they have.

## Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

### CODE REVIEW

File: filebeat/input/udp/input.go [hunk 1]
```diff
@@ -90,9 +104,31 @@ func NewInput(
 		outlet:  out,
 		udp:     udp,
 		started: false,
+		metrics: metrics,
 	}, nil
 }
 
+func configID(config *conf.C) (string, error) {
```
Comment (efd6) [inline filebeat/input/udp/input.go:111]: This is probably a good thing to factor out into inputmon.

Comment (andrewkroh) [inline filebeat/input/udp/input.go:111]: I didn't realize that this input is not converted to the [`v2.Input`](https://github.com/elastic/beats/blob/141ad337a4d28f482e6fe0c97c37ad21d54028e2/filebeat/input/v2/input.go#L60) interface that provides an ID. Can you please scope what the effort would be to convert it to the new interface.

Comment (efd6) [reply filebeat/input/udp/input.go:111]: It looks reasonably straight forward. The only inputs here that are v2 are filestream, journald, kafka and unix. The update would be better in another PR.

File: filebeat/input/udp/input.go [hunk 2]
```diff
@@ -116,10 +152,73 @@ func (p *Input) Stop() {
...
+		packets:        monitoring.NewUint(reg, "udp_packets"),
+		bytes:          monitoring.NewUint(reg, "udp_bytes"),
+		arrivalPeriod:  metrics.NewUniformSample(1024),
+		processingTime: metrics.NewUniformSample(1024),
+	}
+	_ = adapter.NewGoMetrics(reg, "udp_arrival_period", adapter.Accept).
+		Register("histogram", metrics.NewHistogram(out.arrivalPeriod))
+	_ = adapter.NewGoMetrics(reg, "udp_processing_time", adapter.Accept).
+		Register("histogram", metrics.NewHistogram(out.processingTime))
+
```
Comment (andrewkroh) [inline filebeat/input/udp/input.go:198]: Let me propose some slightly more generic names. Before we ship this in the next release I plan to look across the inputs to see if there are opportunities for alignment on naming.

- `received_packets_total`
- `received_events_total` - My thinking is that s/packet/events/ will be more generic and applicable across more inputs.
- `arrival_period`
- `processing_time`

**Review by andrewkroh — COMMENTED** [review]
This looks good. I have some naming suggestions. Can you please a table to the UDP input docs that describes the metrics (the aws-s3 input doc has an example).

File: filebeat/input/udp/input.go (hunk 2 above)
Comment (efd6) [reply filebeat/input/udp/input.go:198]: > - `received_packets_total`
> - `received_events_total` - My thinking is that s/packet/events/ will be more generic and applicable across more 

Is the first of these intended to be `received_bytes_total`?

Comment (andrewkroh) [reply filebeat/input/udp/input.go:198]: Yep. That should have been received_bytes_total.

**Review by andrewkroh — APPROVED** [review]

File: filebeat/input/udp/input.go [hunk 3]
```diff
@@ -116,10 +152,73 @@ func (p *Input) Stop() {
...
+	m.packets.Add(1)
+	m.bytes.Add(uint64(len(data)))
+	if !m.lastPacket.IsZero() {
+		m.arrivalPeriod.Update(timestamp.Sub(m.lastPacket).Nanoseconds())
+	}
+	m.lastPacket = timestamp
+}
+
+func (m *inputMetrics) close() {
+	if m != nil {
```
Comment (efd6) [inline filebeat/input/udp/input.go:220]: Ooops. Will send a fix.

**Pull request merged**